File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/common/data/GameDocument.java
package com.zwstudio.logicpuzzlesandroid.common.data;

import com.j256.ormlite.stmt.DeleteBuilder;
import com.j256.ormlite.stmt.QueryBuilder;
import com.zwstudio.logicpuzzlesandroid.common.domain.Game;
import com.zwstudio.logicpuzzlesandroid.home.android.LogicPuzzlesApplication;

import org.androidannotations.annotations.AfterInject;
import org.androidannotations.annotations.App;
import org.androidannotations.annotations.EBean;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.IOException;
import java.io.InputStream;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import fj.F;
import fj.P;
import fj.P2;

import static fj.data.Array.arrayArray;
import static fj.data.Array.range;
import static fj.data.List.iterableList;

/**
 * Created by zwvista on 2016/09/29.
 */

@EBean(scope = EBean.Scope.Singleton)
public abstract class GameDocument<G extends Game, GM> implements GameDocumentInterface {
    public String gameID() {
        String name = getClass().getSimpleName();
        return name.substring(0, name.indexOf("Document"));
    }

    public List<P2<String, List<String>>> levels = new ArrayList<>();
    public String selectedLevelID;
    public String selectedLevelIDSolution() {return selectedLevelID + " Solution";}
    public List<String> help = new ArrayList<>();
    @App
    public LogicPuzzlesApplication app;

    @AfterInject
    public void init() {
        String filename = gameID() + ".xml";
        try {
            InputStream in_s = app.getApplicationContext().getAssets().open("xml/" + filename);
            loadXml(in_s);
            in_s.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        selectedLevelID = gameProgress().levelID;
    }

    private void loadXml(InputStream in_s) {
        XmlPullParserFactory pullParserFactory;
        try {
            pullParserFactory = XmlPullParserFactory.newInstance();
            XmlPullParser parser = pullParserFactory.newPullParser();

            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
            parser.setInput(in_s, null);

            parseXML(parser);

        } catch (XmlPullParserException e) {

            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void parseXML(XmlPullParser parser) throws XmlPullParserException,IOException
    {
        int eventType = parser.getEventType();
        F<List<String>, List<String>> getCdata = strs -> {
            int a = range(0, strs.size() - 1).find(i -> !strs.get(i).trim().isEmpty()).orSome(-1);
            int b = range(0, strs.size() - 1).reverse().find(i -> !strs.get(i).trim().isEmpty()).orSome(-1);
            return a == -1 || b == -1 ? strs : strs.subList(a, b + 1);
        };
        while (eventType != XmlPullParser.END_DOCUMENT){
            String name = null;
            String id = null;
            List<String> layout = null;
            switch (eventType){
                case XmlPullParser.START_DOCUMENT:
                    break;
                case XmlPullParser.START_TAG:
                    name = parser.getName();
                    if (name.equals("level")){

                        id = parser.getAttributeValue(null,"id");
                        layout = arrayArray(parser.nextText().split("\n"))
                                .map(s -> s.replace("\r", ""))
                                .toJavaList();
                        layout = iterableList(getCdata.f(layout))
                                .map(s -> s.replace("`", ""))
                                .toJavaList();
                        levels.add(P.p(id, layout));
                    } else if (name.equals("help")){
                        help = arrayArray(parser.nextText().split("\n"))
                                .map(s -> s.replace("\r", ""))
                                .toJavaList();
                        help = getCdata.f(help);
                    }
                    break;
                case XmlPullParser.END_TAG:
                    break;
            }
            eventType = parser.next();
        }

    }

    public GameProgress gameProgress() {
        try {
            GameProgress rec = app.daoGameProgress.queryBuilder()
                    .where().eq("gameID", gameID())
                    .queryForFirst();
            if (rec == null) {
                rec = new GameProgress();
                rec.gameID = gameID();
                app.daoGameProgress.create(rec);
            }
            return rec;
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }
    }

    private LevelProgress levelProgress(String levelID) {
        try {
            LevelProgress rec = app.daoLevelProgress.queryBuilder()
                    .where().eq("gameID", gameID())
                    .and().eq("levelID", levelID).queryForFirst();
            if (rec == null) {
                rec = new LevelProgress();
                rec.gameID = gameID();
                rec.levelID = levelID;
                app.daoLevelProgress.create(rec);
            }
            return rec;
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }
    }
    public LevelProgress levelProgress() {
        return levelProgress(selectedLevelID);
    }
    public LevelProgress levelProgressSolution() {
        return levelProgress(selectedLevelIDSolution());
    }

    private List<MoveProgress> moveProgress(String levelID) {
        try {
            QueryBuilder<MoveProgress, Integer> builder = app.daoMoveProgress.queryBuilder();
            builder.where().eq("gameID", gameID())
                    .and().eq("levelID", levelID);
            builder.orderBy("moveIndex", true);
            List<MoveProgress> rec = builder.query();
            return rec;
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }
    }
    public List<MoveProgress> moveProgress() {
        return moveProgress(selectedLevelID);
    }
    public List<MoveProgress> moveProgressSolution() {
        return moveProgress(selectedLevelIDSolution());
    }

    public void levelUpdated(Game game) {
        try {
            LevelProgress rec = levelProgress();
            rec.moveIndex = game.moveIndex();
            app.daoLevelProgress.update(rec);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public void gameSolved(Game game) {
        LevelProgress recLP = levelProgress(), recLPS = levelProgressSolution();
        if (recLPS.moveIndex == 0 || recLPS.moveIndex > recLP.moveIndex)
            saveSolution(game);
    }

    public void moveAdded(Game game, GM move) {
        try {
            DeleteBuilder<MoveProgress, Integer> builder = app.daoMoveProgress.deleteBuilder();
            builder.where().eq("gameID", gameID())
                    .and().eq("levelID", selectedLevelID)
                    .and().ge("moveIndex", game.moveIndex());
            builder.delete();
            MoveProgress rec = new MoveProgress();
            rec.gameID = gameID();
            rec.levelID = selectedLevelID;
            rec.moveIndex = game.moveIndex();
            saveMove(move, rec);
            app.daoMoveProgress.create(rec);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    protected abstract void saveMove(GM move, MoveProgress rec);
    public abstract GM loadMove(MoveProgress rec);

    public void resumeGame() {
        try {
            GameProgress rec = gameProgress();
            rec.levelID = selectedLevelID;
            app.daoGameProgress.update(rec);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public void clearGame() {
        try {
            DeleteBuilder<MoveProgress, Integer> builder = app.daoMoveProgress.deleteBuilder();
            builder.where().eq("gameID", gameID())
                    .and().eq("levelID", selectedLevelID);
            builder.delete();
            LevelProgress rec = levelProgress();
            rec.moveIndex = 0;
            app.daoLevelProgress.update(rec);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    private void copyMoves(List<MoveProgress> moveProgressFrom, String levelIDTo) {
        try {
            DeleteBuilder<MoveProgress, Integer> builder = app.daoMoveProgress.deleteBuilder();
            builder.where().eq("gameID", gameID())
                    .and().eq("levelID", levelIDTo);
            builder.delete();
            for (MoveProgress recMP : moveProgressFrom) {
                GM move = loadMove(recMP);
                MoveProgress recMPS = new MoveProgress();
                recMPS.gameID = gameID();
                recMPS.levelID = levelIDTo;
                recMPS.moveIndex = recMP.moveIndex;
                saveMove(move, recMPS);
                app.daoMoveProgress.create(recMPS);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public void saveSolution(Game game) {
        copyMoves(moveProgress(), selectedLevelIDSolution());
        try {
            LevelProgress rec = levelProgressSolution();
            rec.moveIndex = game.moveIndex();
            app.daoLevelProgress.update(rec);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public void loadSolution() {
        List<MoveProgress> mps = moveProgressSolution();
        copyMoves(mps, selectedLevelID);
        try {
            LevelProgress rec = levelProgress();
            rec.moveIndex = mps.size();
            app.daoLevelProgress.update(rec);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public void deleteSolution() {
        try {
            DeleteBuilder<MoveProgress, Integer> builder = app.daoMoveProgress.deleteBuilder();
            builder.where().eq("gameID", gameID())
                    .and().eq("levelID", selectedLevelIDSolution());
            builder.delete();
            LevelProgress rec = levelProgressSolution();
            rec.moveIndex = 0;
            app.daoLevelProgress.update(rec);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public void resetAllLevels() {
        try {
            DeleteBuilder<MoveProgress, Integer> builder = app.daoMoveProgress.deleteBuilder();
            builder.where().eq("gameID", gameID());
            builder.delete();
            DeleteBuilder<LevelProgress, Integer> builder2 = app.daoLevelProgress.deleteBuilder();
            builder2.where().eq("gameID", gameID());
            builder2.delete();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public int getMarkerOption() {
        String o = gameProgress().option1;
        return o == null ? 0 : Integer.parseInt(o);
    }
    public void setMarkerOption(GameProgress rec, int o) {
        rec.option1 = String.valueOf(o);
    }
    public boolean isAllowedObjectsOnly() {
        String o = gameProgress().option2;
        return Boolean.parseBoolean(o);
    }
    public void setAllowedObjectsOnly(GameProgress rec, boolean o) {
        rec.option2 = String.valueOf(o);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/common/data/GameDocument.kt
package com.zwstudio.logicpuzzlesandroid.common.data

import com.zwstudio.logicpuzzlesandroid.common.domain.Game
import com.zwstudio.logicpuzzlesandroid.home.android.HomeChooseGameActivity
import com.zwstudio.logicpuzzlesandroid.home.android.LogicPuzzlesApplication
import com.zwstudio.logicpuzzlesandroid.home.android.realm
import org.androidannotations.annotations.AfterInject
import org.androidannotations.annotations.App
import org.androidannotations.annotations.EBean
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.InputStream

interface GameDocumentInterface {
    val markerOption: Int
    fun setMarkerOption(rec: GameProgress, o: Int)
    val isAllowedObjectsOnly: Boolean
    fun setAllowedObjectsOnly(rec: GameProgress, o: Boolean)
}

@EBean(scope = EBean.Scope.Singleton)
abstract class GameDocument<GM> : GameDocumentInterface {
    val gameID: String
        get() {
            val name = javaClass.simpleName
            return name.substring(0, name.indexOf("Document"))
        }

    val gameTitle: String
        get() {
            val name = gameID
            return HomeChooseGameActivity.name2title[name] ?: name
        }

    var levels = mutableListOf<GameLevel>()
    lateinit var selectedLevelID: String
    val selectedLevelIDSolution get() = "$selectedLevelID Solution"

    var help = listOf<String>()

    @App
    lateinit var app: LogicPuzzlesApplication

    @AfterInject
    fun init() {
        val filename = gameID + ".xml"
        val in_s = app.applicationContext.assets.open("xml/$filename")
        loadXml(in_s)
        in_s.close()
        selectedLevelID = gameProgress().levelID
    }

    private fun loadXml(in_s: InputStream) {
        val pullParserFactory = XmlPullParserFactory.newInstance()
        val parser = pullParserFactory.newPullParser()
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
        parser.setInput(in_s, null)
        parseXML(parser)
    }

    private fun parseXML(parser: XmlPullParser) {
        fun getCdata(strs: List<String>) = strs.filter { !it.isBlank() }
        var eventType = parser.eventType
        while (eventType != XmlPullParser.END_DOCUMENT) {
            if (eventType == XmlPullParser.START_TAG)
                when (parser.name) {
                    "level" -> {
                        val id = parser.getAttributeValue(null, "id")
                        val settings = mutableMapOf<String, String>()
                        for (i in 0 until parser.attributeCount)
                            settings[parser.getAttributeName(i)] = parser.getAttributeValue(i)
                        var layout = parser.nextText().split("\n").map { it.replace("\r", "") }
                        layout = getCdata(layout).map { it.replace("`", "") }
                        val level = GameLevel()
                        level.id = id
                        level.layout = layout
                        level.settings = settings
                        levels.add(level)
                    }
                    "help" -> {
                        help = parser.nextText().split("\n").map { it.replace("\r", "") }
                        help = getCdata(help)
                    }
                    else -> {}
                }
            eventType = parser.next()
        }
    }

    fun gameProgress(): GameProgress {
        var rec = realm.where(GameProgress::class.java)
            .equalTo("gameID", gameID)
            .findFirst()
        if (rec == null) {
            realm.beginTransaction()
            rec = GameProgress()
            rec.gameID = gameID
            rec.levelID = levels[0].id
            realm.insertOrUpdate(rec)
            realm.commitTransaction()
        }
        return rec
    }

    private fun levelProgress(levelID: String): LevelProgress {
        var rec = realm.where(LevelProgress::class.java)
            .equalTo("gameID", gameID)
            .equalTo("levelID", levelID).findFirst()
        if (rec == null) {
            realm.beginTransaction()
            rec = LevelProgress()
            rec.gameID = gameID
            rec.levelID = levelID
            realm.insertOrUpdate(rec)
            realm.commitTransaction()
        }
        return rec
    }
    fun levelProgress() = levelProgress(selectedLevelID)
    fun levelProgressSolution() = levelProgress(selectedLevelIDSolution)

    private fun moveProgress(levelID: String): List<MoveProgress> =
        realm.where(MoveProgress::class.java)
            .equalTo("gameID", gameID)
            .equalTo("levelID", levelID)
            .sort("moveIndex")
            .findAll()
    fun moveProgress() = moveProgress(selectedLevelID)
    fun moveProgressSolution() = moveProgress(selectedLevelIDSolution)

    fun levelUpdated(game: Game<*, *, *>) {
        realm.beginTransaction()
        val rec = levelProgress()
        rec.moveIndex = game.moveIndex
        realm.insertOrUpdate(rec)
        realm.commitTransaction()
    }

    fun gameSolved(game: Game<*, *, *>) {
        val recLP = levelProgress()
        val recLPS = levelProgressSolution()
        if (recLPS.moveIndex == 0 || recLPS.moveIndex > recLP.moveIndex) saveSolution(game)
    }

    fun moveAdded(game: Game<*, *, *>, move: GM) {
        realm.beginTransaction()
        realm.where(MoveProgress::class.java)
            .equalTo("gameID", gameID)
            .equalTo("levelID", selectedLevelID)
            .greaterThanOrEqualTo("moveIndex", game.moveIndex)
            .findAll().deleteAllFromRealm()
        val rec = MoveProgress()
        rec.gameID = gameID
        rec.levelID = selectedLevelID
        rec.moveIndex = game.moveIndex
        saveMove(move, rec)
        realm.insertOrUpdate(rec)
        realm.commitTransaction()
    }

    protected abstract fun saveMove(move: GM, rec: MoveProgress)
    abstract fun loadMove(rec: MoveProgress): GM
    fun resumeGame() {
        realm.beginTransaction()
        val rec = gameProgress()
        rec.levelID = selectedLevelID
        realm.insertOrUpdate(rec)
        realm.commitTransaction()
    }

    fun clearGame() {
        realm.beginTransaction()
        realm.where(MoveProgress::class.java)
            .equalTo("gameID", gameID)
            .equalTo("levelID", selectedLevelID)
            .findAll().deleteAllFromRealm()
        val rec = levelProgress()
        rec.moveIndex = 0
        realm.insertOrUpdate(rec)
        realm.commitTransaction()
    }

    private fun copyMoves(moveProgressFrom: List<MoveProgress>, levelIDTo: String) {
        realm.where(MoveProgress::class.java)
            .equalTo("gameID", gameID)
            .equalTo("levelID", levelIDTo)
            .findAll().deleteAllFromRealm()
        for (recMP in moveProgressFrom) {
            val move = loadMove(recMP)
            val recMPS = MoveProgress()
            recMPS.gameID = gameID
            recMPS.levelID = levelIDTo
            recMPS.moveIndex = recMP.moveIndex
            saveMove(move, recMPS)
            realm.insertOrUpdate(recMPS)
        }
    }

    fun saveSolution(game: Game<*, *, *>) {
        realm.beginTransaction()
        copyMoves(moveProgress(), selectedLevelIDSolution)
        val rec = levelProgressSolution()
        rec.moveIndex = game.moveIndex
        realm.insertOrUpdate(rec)
        realm.commitTransaction()
    }

    fun loadSolution() {
        realm.beginTransaction()
        val mps = moveProgressSolution()
        copyMoves(mps, selectedLevelID)
        val rec = levelProgress()
        rec.moveIndex = mps.size
        realm.insertOrUpdate(rec)
        realm.commitTransaction()
    }

    fun deleteSolution() {
        realm.beginTransaction()
        realm.where(MoveProgress::class.java)
            .equalTo("gameID", gameID)
            .equalTo("levelID", selectedLevelIDSolution)
            .findAll().deleteAllFromRealm()
        val rec = levelProgressSolution()
        rec.moveIndex = 0
        realm.insertOrUpdate(rec)
        realm.commitTransaction()
    }

    fun resetAllLevels() {
        realm.beginTransaction()
        realm.where(MoveProgress::class.java)
            .equalTo("gameID", gameID)
            .findAll().deleteAllFromRealm()
        realm.where(LevelProgress::class.java)
            .equalTo("gameID", gameID)
            .findAll().deleteAllFromRealm()
        realm.commitTransaction()
    }

    override val markerOption get() = gameProgress().option1?.toInt() ?: 0

    override fun setMarkerOption(rec: GameProgress, o: Int) {
        rec.option1 = o.toString()
    }

    override val isAllowedObjectsOnly get() = gameProgress().option2?.toBoolean() ?: false

    override fun setAllowedObjectsOnly(rec: GameProgress, o: Boolean) {
        rec.option2 = o.toString()
    }
}